(Biometry 2/4) Use biometrics instead of pincode enter screen when enabled

kathaypacific opened the pull request:
### Description

- Create new keychain item that stores pincode protected by biometrics. I opted for using an async function rather than a saga (this was how biometrics was originally implemented) because I think it is nice for the consumer of the function to decide how to handle the outcome. On the EnableBiometry screen, this means staying on the screen while loading/error, and navigating further on success.
- For flows that require pincode, if pincodeType is PhoneAuth then try to fetch pincode via keychain and biometrics. Fallback to pincode enter screen.


### Other changes

Remove unused `isSettingPin` from redux state, remove unused setPincode action

### Tested

Manually & unit tests.

### How others should test

Test flows that require pin with biometrics (face id, touch id). Technically you can test this on a simulator but it doesn't actually match your face/fingerprint so it is better to test on a physical device.

Also test with face/fingerprint obscured to see fallback flow.

Flows like sending tokens, seeing recovery phrase, reset pin should all work with biometrics instead of requiring pin.

Please note that this feature is only available for iOS currently.

### Related issues

- Fixes #1684 

### Backwards compatibility

Yes

kathaypacific commented on packages/mobile/src/onboarding/registration/EnableBiometry.tsx:
```diff
@@ -72,10 +77,13 @@ export default function EnableBiometry({ navigation, route }: Props) {
+      dispatch(setPincodeSuccess(PincodeType.PhoneAuth))
+      navigate(choseToRestoreAccount ? Screens.ImportWallet : Screens.VerificationEducationScreen)
+    } catch (error) {
+      Logger.warn(TAG, 'Error enabling biometrics', error)
```
I've chosen `warn` over `error` for most of the logging because the user action of cancelling the biometric authentication (after 2 failed attempts, if the user is wearing a mask or something) would trigger this error. there isn't a list of error codes provided by the keychain lib to exclude this error - there is only a hard coded string error displayed and it seems OS specific

kathaypacific commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -167,6 +170,18 @@ function storePasswordHash(hash: string, account: string) {
+function storePinWithBiometrics(pin: string) {
+  return storeItem({
+    key: STORAGE_KEYS.PIN,
+    value: pin,
```
should we encrypt this value somehow? 🤔

jeanregisser reviewed (commented):
Amazing work! 🔒 🔑 

What happens when the user restores their iPhone account to a new device?
Answering myself: nothing special, they need to restore their Valora account. So we're good on that front 😄 

Can we cover all paths with tests in the new code added in `authentication.ts` so codecov doesn't show any warning about this?
This is a critical part of the app and we should aim for maximum coverage to avoid surprises.
Here it's mostly about testing some more error cases from what I see.

kathaypacific commented on packages/mobile/ios/celo/Info.plist:
```diff
@@ -74,6 +74,8 @@
 	<string>Connecting your camera allows you to scans codes for payments.</string>
 	<key>NSContactsUsageDescription</key>
 	<string>Adding your contacts makes it easy to send and request payments with your friends.</string>
+	<key>NSFaceIDUsageDescription</key>
+	<string>This is required for you to use Face ID to secure your account.</string>
```
ah good one, i totally forgot! i have done now :)

kathaypacific commented on packages/mobile/src/navigator/NavigationService.ts:
```diff
@@ -123,11 +127,20 @@ export async function ensurePincode(): Promise<boolean> {
+      await getPincodeWithBiometrics()
+      return true
+    } catch (error) {
+      Logger.warn(`${TAG}@ensurePincode`, `Retrieve PIN by biometrics error`, error)
```
okay, i've updated to Logger.error and we can always silence it in Sentry/update it later if it gets noisy.

the chain of calls to fetch the pin with biometrics is `getPincodeWithBiometrics` -> `retrieveStoredItem` -> `Keychain.getGenericPassword`. for the user cancelled action in case of failed biometrics auth,`Keychain.getGenericPassword` will reject with a user cancelled error and this is already logged with `Logger.error` in `retrieveStoredItem`. The new error that is thrown from `retrieveStoredItem` is a generic `ErrorMessages.KEYCHAIN_STORAGE_ERROR` so in this function above, we would be sending the generic `ErrorMessages.KEYCHAIN_STORAGE_ERROR` to Sentry on top of the actual keychain error sent by `retrieveStoredItem`.....my thinking is that most errors that come from this event will be user cancelled errors but let's see in Sentry :)

kathaypacific commented on packages/mobile/src/pincode/PincodeSet.tsx:
```diff
@@ -71,6 +72,7 @@ function mapStateToProps(state: RootState): StateProps {
 const mapDispatchToProps = {
   setPincode,
```
good catch! not anymore, i've removed this action and the `isSettingPin` state that wasn't used as well :D

kathaypacific commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -252,6 +267,43 @@ export function* getPasswordSaga(account: string, withVerification?: boolean, st
 
 type PinCallback = (pin: string) => void
 
+export async function setPincodeWithBiometrics() {
+  let pin = getCachedPin(DEFAULT_CACHE_ACCOUNT)
+  if (!pin) {
+    pin = await requestPincodeInput(true, true)
+  }
```
i'm not sure what you mean here? i really wanted `setPincodeWithBiometrics` to be simple to call, because this will be called from the biometry screen + the settings screen. i didn't want those screens to know about how to fetch the pin. there is a `storePinWithBiometrics` function a few lines down that does explicitly require the pin as a param, is this what you mean?

jeanregisser replied:
Ah yes I see, fine as is then 👍

kathaypacific commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -65,8 +75,23 @@ describe(getPincode, () => {
+    await getPincode()
+
+    expect(getPasswordSpy).toHaveBeenCalledTimes(1)
+    expect(getPasswordSpy).toHaveBeenCalledWith({ service: 'PIN' })
```
🙈 because i didn't know you could do that 🙈 i've updated, i'm happy to have learned this!

kathaypacific commented:
> Amazing work! 🔒 🔑
> 
> What happens when the user restores their iPhone account to a new device? Answering myself: nothing special, they need to restore their Valora account. So we're good on that front 😄
> 
> Can we cover all paths with tests in the new code added in `authentication.ts` so codecov doesn't show any warning about this? This is a critical part of the app and we should aim for maximum coverage to avoid surprises. Here it's mostly about testing some more error cases from what I see.

Actually really happy that you suggested to do more testing, in the process i found that the `getPasswordSaga` had a global mock so we were testing a mock 🥲

kathaypacific commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -82,6 +163,174 @@ describe(getPincode, () => {
+    )
+    expect(mockedKeychain.setGenericPassword).toHaveBeenCalledWith(
+      'CELO',
+      expect.any(String), // TODO test that this can be decrypted correctly
```
@jeanregisser any suggestions on how to test that the correct encrypted mnemonic is stored? the encryption doesn't resolve to a single value so i couldn't hard code this, and we're mocking the keychain so I couldn't retrieve it to test it 🥲

jeanregisser replied:
Since it's a unit test, it's fine if it's not entirely real.
But can we have a deterministic string here?

No need to do it here, but testing this e2e would ensure everything is really working as expected.

gnardini commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -252,6 +267,43 @@ export function* getPasswordSaga(account: string, withVerification?: boolean, st
+      throw new Error('Retrieved incorrect pin with biometrics after saving')
+    }
+  } catch (error) {
+    Logger.error(TAG, 'Failed to save pin with biometrics', error)
```
Yes!

gnardini commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -319,8 +381,12 @@ export async function updatePin(account: string, oldPin: string, newPin: string)
+      const hash = getPasswordHash(newPassword)
       await storePasswordHash(hash, account)
+      const pincodeType = pincodeTypeSelector(store.getState())
+      if (pincodeType === PincodeType.PhoneAuth) {
```
I'm _a bit_ worried about what happens if an intermediate steps fail, but I'm not sure if the worry makes sense or not 🤔  We have a blanket error catch, but say we save the PIN this way but then updating the mnemonic fails. I guess that's where the asking for PIN fallback comes in 😄 

Yeah, writing this I think I convinced myself that it's fine, but sending the message anyways just in case

kathaypacific replied:
yeah indeed, i've tried to make the phone auth as "optional" as possible - basically just try and if it fails then the user just gets the flow they currently have

gnardini reviewed (commented):
Amazing job! Can't wait to see this live 😄

kathaypacific commented on packages/mobile/src/navigator/NavigationService.ts:
```diff
@@ -123,11 +127,20 @@ export async function ensurePincode(): Promise<boolean> {
+      await getPincodeWithBiometry()
+      return true
+    } catch (error) {
+      Logger.error(`${TAG}@ensurePincode`, `Retrieve PIN by biometry error`, error)
```
i've added a comment explaining! we need the flow to continue since we do not provide a fallback for the biometric auth - if face/touch id fails, then we need to continue in the function to activate the pincode input screen :)

kathaypacific commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -252,6 +267,43 @@ export function* getPasswordSaga(account: string, withVerification?: boolean, st
+      throw new Error('Retrieved incorrect pin with biometry after saving')
+    }
+  } catch (error) {
+    Logger.error(TAG, 'Failed to save pin with biometry', error)
```
i've updated :D

kathaypacific commented on packages/mobile/src/pincode/authentication.ts:
```diff
@@ -260,6 +312,16 @@ export async function getPincode(withVerification = true) {
+      const retrievedPin = await getPincodeWithBiometry()
+      return retrievedPin
+    } catch (error) {
+      Logger.error(TAG, 'Failed to retrieve pin with biometry', error)
```
ditto, i've added the comment here as well explaining the pin input fallback :)

gnardini reviewed (approved):
Amazing job!

jeanregisser commented on packages/mobile/src/navigator/NavigationService.ts:
```diff
@@ -134,10 +134,12 @@ export async function ensurePincode(): Promise<boolean> {
       return true
     } catch (error) {
-      Logger.warn(`${TAG}@ensurePincode`, `Retrieve PIN by biometrics error`, error)
+      Logger.error(`${TAG}@ensurePincode`, `Retrieve PIN by biometry error`, error)
```
I saw Sentry captured an error about user cancelling this: https://sentry.io/organizations/valora-inc/issues/2956635589/?environment=org.celo.mobile.alfajores.dev&project=1250733&query=canceled&statsPeriod=14d
We should check that here and just warn for that case.

kathaypacific replied:
yeah this was the issue i mentioned above, where the keychain lib throws an error for the user cancelled event and we were catching it in `retrieveStoredItem` and throwing a generic `throw new Error(ErrorMessages.KEYCHAIN_STORAGE_ERROR)`

i've attempted to exclude the user cancelled error from the genuine errors the same way as we exclude network errors, would be nice to see if you agree with the handling :)

jeanregisser commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -174,13 +254,92 @@ describe(setPincodeWithBiometrics, () => {
+  const expectedPasswordHash = '9853810edb88b031bf6ac1505f5689cb423876fbeb14f7a3037c97ec4531b6ae'
+  // expectedAccountHash generated from normalizeAddress(mockAccount)
+  const expectedAccountHash = 'PASSWORD_HASH-0000000000000000000000000000000000007e57'
+  // mockEncryptedMneumonic generated using mockMneumonic, mockAccount, old pin
+  const mockEncryptedMneumonic =
```
```suggestion
  // mockEncryptedMnemonic generated using mockMnemonic, mockAccount, old pin
  const mockEncryptedMnemonic =
```

kathaypacific replied:
haha good catch. i've updated!

jeanregisser reviewed (commented):
Great job covering this + existing things with more tests! 👍 

Left a couple of small comments.

kathaypacific commented:
@jeanregisser to your comment above 

> Since it's a unit test, it's fine if it's not entirely real.
> But can we have a deterministic string here?
> 
> No need to do it here, but testing this e2e would ensure everything is really working as expected.

I was having issues making the encryption deterministic before, but ended up mocking the crypto-js lib and checking that it was called correctly :)

jeanregisser commented on packages/mobile/src/navigator/NavigationService.ts:
```diff
@@ -123,11 +128,24 @@ export async function ensurePincode(): Promise<boolean> {
+      await getPincodeWithBiometry()
+      return true
+    } catch (error) {
+      if (!isUserCancelledError(error)) {
+        Logger.error(`${TAG}@ensurePincode`, `Retrieve PIN by biometry error`, error)
+      }
```
Great!

jeanregisser commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -7,45 +8,92 @@ import { navigate, navigateBack } from 'src/navigator/NavigationService'
+    expectPincodeEntered()
+  })
+  it('should throw an error for unset pincode type', async () => {
+    try {
+      await expectSaga(getPasswordSaga, mockAccount, false, false)
+        .provide([[select(pincodeTypeSelector), PincodeType.Unset]])
+        .run()
+    } catch (error) {
+      expect(error).toEqual(Error('Pin has never been set'))
+    }
```
This will happily pass if no error is actually thrown.
Related to our discussion in https://github.com/valora-inc/wallet/pull/1816#discussion_r793482703

kathaypacific replied:
✅ i've updated all these already in https://github.com/valora-inc/wallet/pull/1816

jeanregisser commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -7,45 +8,92 @@ import { navigate, navigateBack } from 'src/navigator/NavigationService'
+    }
+  })
+  it('should throw an error for unexpected pincode type', async () => {
+    try {
+      await expectSaga(getPasswordSaga, mockAccount, false, false)
+        .provide([[select(pincodeTypeSelector), 'unexpectedPinType']])
+        .run()
+    } catch (error) {
+      expect(error).toEqual(Error('Unsupported Pincode Type unexpectedPinType'))
+    }
```
ditto

jeanregisser commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -82,6 +180,186 @@ describe(getPincode, () => {
+  it('throws an error if a null pin was retrieved', async () => {
+    mockedKeychain.getGenericPassword.mockResolvedValue(false)
+
+    try {
+      await getPincodeWithBiometry()
+    } catch (error) {
+      expect(error).toEqual(expect.any(Error))
+    }
```
ditto

jeanregisser commented on packages/mobile/src/pincode/authentication.test.ts:
```diff
@@ -82,6 +180,186 @@ describe(getPincode, () => {
+      storage: 'storage',
+    })
+
+    try {
+      await setPincodeWithBiometry()
+    } catch (error) {
+      expect(error).toEqual(expect.any(Error))
+    }
```
ditto

jeanregisser commented on packages/mobile/src/storage/keychain.tsx:
```diff
@@ -36,6 +46,9 @@ export async function retrieveStoredItem(key: string) {
     }
     return item.password
   } catch (error) {
+    if (isUserCancelledError(error)) {
+      throw error
+    }
```
Maybe we should add a comment here why we're still re-throwing when cancelled?

kathaypacific replied:
done! i hope it's clear from the comment..but i think it is still necessary/easier for the consumer of this function to be aware that fetching from the keychain failed (to know whether to log an error, to default to the pin input screen etc) - i toyed with the idea of returning null instead of throwing an error in this case, but then consumers need to know that null value means either the user cancelled or the keychain returned a falsey value :/ this seemed like the worse option

kathaypacific replied:
i thought about this in the weekend, i don't think it should return an error....i will change this

kathaypacific commented:
@jeanregisser ready for another pass:
- i removed the keychain store/retrieve/remove functions to throw the actual error instead of ErrorMessages.KeychainError. i think this is less confusing than throwing a generic error + a real error for the user cancelled case....and ErrorMessages.KeychainError is not currently being used so i am confident i haven't modified any existing behaviours.
- I didn't remove KeychainError from ErrorMessages, i think for a future PR we should evaluate where the keychain errors are unhandled and display a generic keychain error to the user. (e.g. in the pincode screens where we use `checkPin` we should catch the error)
- since the Keychain functions are already using Logger.error (for non user cancelled errors), i demoted the new biometrics logging that use the keychain functions to `warn` to avoid sending the same error to sentry twice
- i moved the test assertions changes in #1816 to here to make it less confusing

jeanregisser reviewed (approved):
🎉

kathaypacific commented:
@mergifyio refresh

ValoraQA commented:
Hey @MuckT, We have verified the above task in Latest Test flight Alfajores build V1.26.0 and updated the status in test rails.
**Note:**

- We have verified face ID for iOS 14,15 & touch ID for iOS13
- As this features is iOS specific, so we have blocked android test cases

**Test rail links:**

**iOS 13:** https://valoraapp.testrail.io/index.php?/tests/view/9999

**iOS 14:** https://valoraapp.testrail.io/index.php?/tests/view/9744

**iOS 15:** https://valoraapp.testrail.io/index.php?/tests/view/9681

**Devices:** iPhone 7 (13.3), iPhone 13 mini (15.1.1), iPhone 12 (14.7.1)

**CC:** @kathaypacific 

Thanks..!!
